### PR TITLE
Void checks

### PR DESCRIPTION
**Pull request by a14n**
Fixes dart-lang/sdk#32827

### CODE REVIEW

File: test/rules/void_checks.dart [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+//   f<T>(T Function() f) => f();
+//   f(() // OK
+//       {
+//     return 1;
+//   });
+//   f<void>(() // LINT
+//       {
+//     return 1;
+//   });
+// }
```
Comment (a14n) [inline test/rules/void_checks.dart]: I really need some help for those tests. I didn't manage to handle generics correctly :-(

**Comment by a14n** [comment]
It looks like I made some mistakes for return in `MethodInvocation`. I will make some changes.

**Comment by a14n** [comment]
It's ready for review. I run it on Flutter without errors.

File: lib/src/rules/void_checks.dart [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    final args = node.argumentList.arguments;
+    final parameters = node.staticElement.parameters;
+    _checkArgs(args, parameters);
+  }
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    final type = node.leftHandSide?.bestType;
+    if (!_isFunctionRef(type, node.rightHandSide)) {
+      _check(node.leftHandSide?.bestType, node.rightHandSide?.bestType, node);
```
Comment (bwilkerson) [inline lib/src/rules/void_checks.dart]: "node.leftHandSide?.bestType" --> "type"

File: lib/src/rules/void_checks.dart [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    } else if (parent is FunctionDeclaration) {
+      _check(parent.element.returnType, node.expression?.bestType, node);
+    }
+  }
+
+  void _checkArgs(
+      NodeList<Expression> args, List<ParameterElement> parameters) {
+    final positionalParameters = parameters.where((e) => !e.isNamed).toList();
+    int positionalCount = 0;
+    for (final arg in args) {
```
Comment (bwilkerson) [inline lib/src/rules/void_checks.dart]: We might be able to simplify this code by using `Expression.bestParameterElement`.

File: lib/src/rules/void_checks.dart [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  bool _isFunctionRef(DartType type, Expression arg) =>
+      type is FunctionType &&
+      (arg is SimpleIdentifier ||
+          arg is PrefixedIdentifier ||
+          arg is FunctionExpression && arg.body is ExpressionFunctionBody);
+
+  void _check(DartType expectedType, DartType type, AstNode node) {
+    if (expectedType == null || type == null)
```
Comment (bwilkerson) [inline lib/src/rules/void_checks.dart]: nit: please add a block around the `return;`.

File: lib/src/rules/void_checks.dart (hunk 2 above)
Comment (a14n) [reply lib/src/rules/void_checks.dart]: done

File: lib/src/rules/void_checks.dart (hunk 3 above)
Comment (a14n) [reply lib/src/rules/void_checks.dart]: done

File: lib/src/rules/void_checks.dart (hunk 4 above)
Comment (a14n) [reply lib/src/rules/void_checks.dart]: done

**Pull request merged**